File: lib/core/reader/epub/epub_archive_parser.dart
import 'dart:io';

import 'package:archive/archive.dart';

class ParsedEpub {
  final String htmlContent;
  final int chapterCount;

  const ParsedEpub({required this.htmlContent, required this.chapterCount});
}

class EpubArchiveParser {
  ParsedEpub parseCombinedHtml(String epubPath) {
    final file = File(epubPath);
    if (!file.existsSync()) {
      throw FileSystemException('EPUB file missing', epubPath);
    }

    final bytes = file.readAsBytesSync();
    final archive = ZipDecoder().decodeBytes(bytes);

    final rootOpfPath = _findRootPackagePath(archive);
    final opfContent = _readText(archive, rootOpfPath);
    final baseDir = rootOpfPath.contains('/')
        ? rootOpfPath.substring(0, rootOpfPath.lastIndexOf('/'))
        : '';

    final manifest = _parseManifest(opfContent);
    final spineItemIds = _parseSpine(opfContent);

    final chapters = <String>[];
    for (final itemId in spineItemIds) {
      final href = manifest[itemId];
      if (href == null) continue;
      final entryPath = _resolveRelativePath(baseDir, href);
      final text = _readTextOrNull(archive, entryPath);
      if (text != null) chapters.add(text);
    }

    if (chapters.isEmpty) {
      throw const FormatException('No readable EPUB chapters found');
    }

    final buffer = StringBuffer()
      ..write('<html><head><meta charset="utf-8"/></head><body>');
    for (var i = 0; i < chapters.length; i++) {
      buffer
        ..write('<section data-chapter="')
        ..write(i)
        ..write('">')
        ..write(_stripHtmlEnvelope(chapters[i]))
        ..write('</section>');
    }
    buffer.write('</body></html>');

    return ParsedEpub(
      htmlContent: buffer.toString(),
      chapterCount: chapters.length,
    );
  }

  String _findRootPackagePath(Archive archive) {
    final containerFile = archive.findFile('META-INF/container.xml');
    if (containerFile == null) {
      throw const FormatException('Invalid EPUB: missing META-INF/container.xml');
    }
    final containerXml = String.fromCharCodes(containerFile.content as List<int>);

    final match =
        RegExp(r'full-path\s*=\s*"([^"]+)"').firstMatch(containerXml);
    if (match == null) {
      throw const FormatException('Invalid EPUB: OPF package path not found');
    }
    return match.group(1)!;
  }

  Map<String, String> _parseManifest(String opfContent) {
    final regex =
        RegExp(r'<item[^>]*id\s*=\s*"([^"]+)"[^>]*href\s*=\s*"([^"]+)"[^>]*/?>');
    final map = <String, String>{};
    for (final match in regex.allMatches(opfContent)) {
      map[match.group(1)!] = match.group(2)!;
    }
    return map;
  }

  List<String> _parseSpine(String opfContent) {
    final regex = RegExp(r'<itemref[^>]*idref\s*=\s*"([^"]+)"[^>]*/?>');
    return regex.allMatches(opfContent).map((m) => m.group(1)!).toList();
  }

  String _resolveRelativePath(String baseDir, String href) {
    if (href.startsWith('/')) return href.substring(1);
    final cleanHref = href.contains('#') ? href.substring(0, href.indexOf('#')) : href;
    return baseDir.isEmpty ? cleanHref : '$baseDir/$cleanHref';
  }

  String _stripHtmlEnvelope(String rawChapter) {
    final match =
        RegExp(r'<body[^>]*>([\s\S]*?)</body>', caseSensitive: false)
            .firstMatch(rawChapter);
    return match?.group(1) ?? rawChapter;
  }

  String _readText(Archive archive, String path) {
    final file = archive.findFile(path);
    if (file == null) throw FormatException('Missing EPUB entry: $path');
    return String.fromCharCodes(file.content as List<int>);
  }

  String? _readTextOrNull(Archive archive, String path) {
    final file = archive.findFile(path);
    if (file == null) return null;
    return String.fromCharCodes(file.content as List<int>);
  }
}

File: core/reader/src/main/kotlin/com/readler/core/reader/epub/EpubArchiveParser.kt
package com.readler.core.reader.epub

import java.io.File
import java.io.IOException
import java.util.zip.ZipEntry
import java.util.zip.ZipFile

internal class EpubArchiveParser {
    fun parseCombinedHtml(epubPath: String): ParsedEpub {
        val file = File(epubPath)
        require(file.exists()) { "EPUB file missing: $epubPath" }

        ZipFile(file).use { zip ->
            val rootOpfPath = findRootPackagePath(zip)
            val opfContent = zip.readText(rootOpfPath)
            val baseDir = rootOpfPath.substringBeforeLast('/', missingDelimiterValue = "")

            val manifest = parseManifest(opfContent)
            val spineItemIds = parseSpine(opfContent)

            val chapters = spineItemIds.mapNotNull { itemId ->
                val href = manifest[itemId] ?: return@mapNotNull null
                val entryPath = resolveRelativePath(baseDir, href)
                zip.readTextOrNull(entryPath)
            }

            if (chapters.isEmpty()) {
                throw IOException("No readable EPUB chapters found in $epubPath")
            }

            val html = buildString {
                append("<html><head><meta charset=\"utf-8\"/></head><body>")
                chapters.forEachIndexed { index, chapter ->
                    append("<section data-chapter=\"")
                    append(index)
                    append("\">")
                    append(stripHtmlEnvelope(chapter))
                    append("</section>")
                }
                append("</body></html>")
            }

            return ParsedEpub(htmlContent = html, chapterCount = chapters.size)
        }
    }

    private fun findRootPackagePath(zip: ZipFile): String {
        val containerEntry = zip.getEntry("META-INF/container.xml")
            ?: throw IOException("Invalid EPUB: missing META-INF/container.xml")
        val containerXml = zip.readText(containerEntry)

        val match = Regex("""full-path\s*=\s*\"([^\"]+)\"""")
            .find(containerXml)
            ?.groupValues
            ?.getOrNull(1)

        return requireNotNull(match) { "Invalid EPUB: OPF package path not found" }
    }

    private fun parseManifest(opfContent: String): Map<String, String> {
        val regex = Regex("""<item[^>]*id\s*=\s*\"([^\"]+)\"[^>]*href\s*=\s*\"([^\"]+)\"[^>]*/?>""")
        return regex.findAll(opfContent).associate { match ->
            val id = match.groupValues[1]
            val href = match.groupValues[2]
            id to href
        }
    }

    private fun parseSpine(opfContent: String): List<String> {
        val regex = Regex("""<itemref[^>]*idref\s*=\s*\"([^\"]+)\"[^>]*/?>""")
        return regex.findAll(opfContent).map { it.groupValues[1] }.toList()
    }

    private fun resolveRelativePath(baseDir: String, href: String): String {
        if (href.startsWith('/')) {
            return href.removePrefix("/")
        }

        return if (baseDir.isBlank()) {
            href.substringBefore('#')
        } else {
            "$baseDir/${href.substringBefore('#')}"
        }
    }

    private fun stripHtmlEnvelope(rawChapter: String): String {
        val bodyMatch = Regex("""<body[^>]*>([\s\S]*?)</body>""", RegexOption.IGNORE_CASE)
            .find(rawChapter)
            ?.groupValues
            ?.getOrNull(1)
        return bodyMatch ?: rawChapter
    }

    private fun ZipFile.readText(path: String): String {
        val entry = getEntry(path) ?: throw IOException("Missing EPUB entry: $path")
        return readText(entry)
    }

    private fun ZipFile.readTextOrNull(path: String): String? {
        val entry = getEntry(path) ?: return null
        return readText(entry)
    }

    private fun ZipFile.readText(entry: ZipEntry): String {
        return getInputStream(entry).bufferedReader().use { it.readText() }
    }
}

internal data class ParsedEpub(
    val htmlContent: String,
    val chapterCount: Int
)
